Text not wrapping in nested grid

stwrt opened the pull request:
#### 705acfb695e5f37fa7e52c7d2050e9e3151cb606
<pre>
Text not wrapping in nested grid
<a href="https://bugs.webkit.org/show_bug.cgi?id=254214">https://bugs.webkit.org/show_bug.cgi?id=254214</a>
rdar://107002717

Reviewed by Alan Baradlay.

This is a hack fix to support not re-layout text in inner grids. This is caused by the updateGridAreaLogicalSize
being called in performGridItemsPreLayout(). This causes the paragraph to re-need a layout, which should not be happening.
This re-layout ultimately the text to go out of the bounds of the grid.

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-relayout-with-nested-grid-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-relayout-with-nested-grid-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-relayout-with-nested-grid.html: Added.
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::layoutGrid):
(WebCore::RenderGrid::layoutMasonry):
(WebCore::RenderGrid::computeIntrinsicLogicalWidths const):
(WebCore::RenderGrid::performGridItemsPreLayout const):
* Source/WebCore/rendering/RenderGrid.h:

Canonical link: <a href="https://commits.webkit.org/264252@main">https://commits.webkit.org/264252@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6d7331eefd27575e1ca1d2f9d0e624c94f17cec5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7141 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7389 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7567 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8758 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7362 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7151 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8657 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7321 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10263 "101 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7268 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7929 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6577 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8864 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/5283 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6496 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14240 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6944 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6595 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/9473 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7077 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5786 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6437 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/6405 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1696 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10636 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6820 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/fast/css-grid-layout/text-resizing-overflow-expected.txt:
```diff
@@ -0,0 +1,13 @@
+layer at (0,0) size 800x600
```
If the issue is a missing layout, can we test this by having a ref that performs a layout, and a test that does not perform a layout ?

nt1m replied:
though it might be possible to make a more straightforward ref that isn't a render tree dump

alanbaradlay replied:
yeah, or just come up with something that looks exactly like the correct rendering. Render tree dumps have huge maintenance burden and for that reason we stopped producing them for a while now.

stwrt replied:
Switched to HTML expected.

smfr commented on LayoutTests/fast/css-grid-layout/text-resizing-overflow.html:
```diff
@@ -0,0 +1,36 @@
+<script>
+/* Force a re-layout */
+async function forceLayout() {
+  await new Promise(res => setTimeout(res, 50));
```
Why the 50ms timer?

brentfulgham replied:
Yes, we generally only want '0' timeouts (if you need to do something on the next runloop). Anything else is begging for flaky tests.

stwrt replied:
Converted to 0ms timer.

smfr commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -825,6 +827,11 @@ LayoutUnit GridTrackSizingAlgorithmStrategy::logicalHeightForChild(RenderBox& ch
+
+        if (is<RenderBox>(child.firstChild())) {
+            for (auto itr = RenderIterator<RenderBox>(&child, child.firstChildBox()); itr; itr = itr.traverseNext())
+                itr->setNeedsLayout(MarkOnlyThis);
```
Why is it OK for a `const` getter called `logicalHeightForChild` to set children as needing layout?

brentfulgham replied:
@smfr: I'm not sure either, but note that existing code already does this, so perhaps this is expected here.

brentfulgham replied:
@stwrt Do we not have a visitor thing that knows how to walk through the children and mark them as needing layout? I vaguely remember this existing, but I could be wrong.

brentfulgham replied:
I think we might want something like the following (though it looks like other render code uses your approach, so it might be preferable):

```
 for (auto& child : descendantsOfType<RenderBox>(child.firstChild()))
         child.setNeedsLayout(MarkOnlyThis);
```

alanbaradlay replied:
This could should use either an explicit child or descendant iterator.

alanbaradlay replied:
Also this fix looks very specific to the case where the nested grid is a second level child of the grid item. I assume dirty bits need to be propagated further down in case of deeper block level structure.

alanbaradlay replied:
> I think we might want something like the following (though it looks like other render code uses your approach, so it might be preferable):
> 
> ```
>  for (auto& child : descendantsOfType<RenderBox>(child.firstChild()))
>          child.setNeedsLayout(MarkOnlyThis);
> ```
descendant iterator may dirty more that what we need (as it traverses the entire subtree). The issue here is that marking only the direct children (grid items) is not sufficient but the same time marking the entire subtree may be an overkill. In theory such invalidation should stop at formatting context boundary, but we don't have such functionality in render tree yet.

alanbaradlay replied:
I started a debugging session on this to be able to provide a (semi)educated review feedback..but it should not stop others from reviewing it (and landing) if urgent.

brentfulgham reviewed (changes requested):
Please revise the test to not use a hard-coded timeout.

Also consider making a WPT-style test (and promoting this new test to the WPT project).

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-relayout-with-nested-grid.html:
```diff
@@ -0,0 +1,36 @@
+forceLayout();
+</script>
+
+</html>
```
nit: orphan end tag for `<html>`

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-relayout-with-nested-grid-expected.html:
```diff
@@ -0,0 +1,22 @@
+</div>
+</body>
+
+</html>
```
nit: orphan end tag for `<html>`

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-relayout-with-nested-grid.html:
```diff
@@ -0,0 +1,45 @@
+  document.getElementById("extra_div1").appendChild(document.createElement("p"));
+};
+
+forceLayout();
```
`forceLayout()` is async, so you want to add a `class="reftest-wait"` to the `<html>`, and remove the class once the layout is done. Otherwise, the screenshot will be taken before.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-relayout-with-nested-grid.html:
```diff
@@ -0,0 +1,41 @@
+<script>
+/* Force a re-layout */
+document.getElementById("extra_div1").appendChild(document.createElement("p"));
+document.documentElement.classList.remove("reftest-wait");
```
reftest-wait shouldn't be necessary with the setTimeout removed. It's only useful for async operations. I assume you want the setTimeout() so this layout doesn't count as part of the initial one?

stwrt replied:
Correct, we should be doing a second layout, so I will need to to put in the `setTimeout()` again.

alanbaradlay commented on Source/WebCore/rendering/RenderGrid.h:
```diff
@@ -158,7 +158,7 @@ class RenderGrid final : public RenderBlock {
     std::unique_ptr<OrderedTrackIndexSet> computeEmptyTracksForAutoRepeat(GridTrackSizingDirection) const;
 
-    void performGridItemsPreLayout(const GridTrackSizingAlgorithm&) const;
+    void performGridItemsPreLayout(const GridTrackSizingAlgorithm&, bool shouldUpdateGridAreaLogicalSize) const;
```
please change this to enum class.  boolean blindness is real.

stwrt committed:
Text not wrapping in nested grid
https://bugs.webkit.org/show_bug.cgi?id=254214
rdar://107002717

Reviewed by Alan Baradlay.

This is a hack fix to support not re-layout text in inner grids. This is caused by the updateGridAreaLogicalSize
being called in performGridItemsPreLayout(). This causes the paragraph to re-need a layout, which should not be happening.
This re-layout ultimately the text to go out of the bounds of the grid.

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-relayout-with-nested-grid-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-relayout-with-nested-grid-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-relayout-with-nested-grid.html: Added.
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::layoutGrid):
(WebCore::RenderGrid::layoutMasonry):
(WebCore::RenderGrid::computeIntrinsicLogicalWidths const):
(WebCore::RenderGrid::performGridItemsPreLayout const):
* Source/WebCore/rendering/RenderGrid.h:

Canonical link: https://commits.webkit.org/264252@main
